### PR TITLE
Convert to use typesafe features in data types

### PR DESCRIPTION
**Pull request by newhoggy**
# Description

Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue.

# Changelog

```yaml
- description: |
    <insert-changelog-description-here>
  # no-changes: the API has not changed
  # compatible: the API has changed but is non-breaking
  # breaking: the API has changed in a breaking way
  compatibility: <no-api-changes|compatible|breaking>
  # feature: the change implements a new feature in the API
  # bugfix: the change fixes a bug in the API
  # test: the change fixes modifies tests
  # maintenance: the change involves something other than the API
  # If more than one is applicable, it may be put into a list.
  type: <feature|bugfix|test|maintenance>
```

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Feature.hs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    -> CardanoEra era
+    -> a
+
+supportedInShelleyBasedEra :: ()
+  => Feature feature
+  => a
+  -> (feature era -> a)
+  -> ShelleyBasedEra era
+  -> a
+supportedInShelleyBasedEra no yes = supportedInEra no yes . shelleyBasedToCardanoEra
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Feature.hs:33]: Why `no` then `yes`?

File: cardano-api/internal/Cardano/Api/TxBody.hs [hunk 2]
```diff
@@ -300,58 +297,26 @@ isValidToScriptValidity :: L.IsValid -> ScriptValidity
...
-
-scriptValidityToTxScriptValidity :: ShelleyBasedEra era -> ScriptValidity -> TxScriptValidity era
-scriptValidityToTxScriptValidity era scriptValidity = case txScriptValiditySupportedInShelleyBasedEra era of
-  Nothing -> TxScriptValidityNone
-  Just witness -> TxScriptValidity witness scriptValidity
-
-txScriptValidityToIsValid :: TxScriptValidity era -> L.IsValid
-txScriptValidityToIsValid = scriptValidityToIsValid . txScriptValidityToScriptValidity
+data TxScriptValidityFeature era where
+  TxScriptValiditySupportedInAlonzoEra  :: TxScriptValidityFeature AlonzoEra
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/TxBody.hs:301]: [Why do we need to have a constructor for each pair feature-era?](https://github.com/input-output-hk/cardano-api/pull/4#discussion_r1213334225)

File: cardano-api/internal/Cardano/Api/Feature.hs
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Feature.hs:1]: Can you add haddocks?

File: cardano-api/internal/Cardano/Api/Feature.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Feature.hs:33]: Standard Haskell ordering: https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Bool.html#v:bool

This is generally true of all just functions.  For example `maybe` and `either` are functions that put the negative case first.

File: cardano-api/internal/Cardano/Api/TxBody.hs (hunk 2 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/TxBody.hs:301]: Note, I've not changed the code.  I've only renamed `TxScriptValiditySupportedInEra` to `TxScriptValidityFeature`.

The constructor serves as a witness that the pair is valid.  If someone wants to construct a `FeatureValue` for a particular feature/era pair they must supply the witness.  The witness proves the pair is valid and if one cannot be supplied, that means the pair is invalid.

This is how invalid pairs are made into compile errors making the code typesafe.

File: cardano-api/internal/Cardano/Api/Feature.hs
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Feature.hs:1]: I plan to add haddocks as soon as I verify the solution is viable for protocol parameters as well.

**Review by carbolymer — APPROVED** [review]

**Pull request closed without merging**